autopilot/followpoint.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ *     Copyright 2024 Rickard Häll   rickard.hall@ri.se
+ *     Published under GPLv3: https://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+#include "followpoint.h"
+#include "core/geometry.h"
+#include <QDebug>
+#include <QLineF>
+
+FollowPoint::FollowPoint(QSharedPointer<MovementController> movementController)
+{
+    mMovementController = movementController;
+    connect(&mUpdateStateTimer, &QTimer::timeout, this, &FollowPoint::updateState);
+
+    // ToDo: Provide system parameters to ControlTower
+
+    // Follow point requires continuous updates of the point to follow
+    mFollowPointTimedOut = true;
+    FollowPoint::mFollowPointHeartbeatTimer.setSingleShot(true);
+    connect(&mFollowPointHeartbeatTimer, &QTimer::timeout, this, [&](){
+        if ((mStmState == FollowPointSTMstates::FOLLOWING || mStmState == FollowPointSTMstates::WAITING) && this->isActive()) {
+            qDebug() << "WARNING: Follow point timed out. Stopping WaypointFollower.";
+            this->stopFollowPoint();
+        }
+        mFollowPointTimedOut = true;
+    });
+}
+
+FollowPoint::FollowPoint(QSharedPointer<VehicleConnection> vehicleConnection, PosType posTypeUsed)
+{
+    mVehicleConnection = vehicleConnection;
+    connect(&mUpdateStateTimer, &QTimer::timeout, this, &FollowPoint::updateState);
+    setPosTypeUsed(posTypeUsed);
+
+    // TODO: follow point over remote connection not supported for now
+}
+
+bool FollowPoint::isActive()
+{
+    return mUpdateStateTimer.isActive();
+}
+
+void FollowPoint::startFollowPoint()
+{
+    // Deactivate emergency brake in order to use follow point
+    emit deactivateEmergencyBrake();
+    if (isOnVehicle()) {
+        mStmState = FollowPointSTMstates::FOLLOWING;
+        mUpdateStateTimer.start(mUpdateStatePeriod_ms);
+    } else {
+        if (isOnVehicle())
+            qDebug() << "WARNING: Follow Point did not get a recent point to follow. Exiting.";
+        else
+            qDebug() << "WARNING: Follow Point not implemented for remote connections. Exiting.";
+        mStmState = FollowPointSTMstates::STOP;
```

Comment:
Would add mUpdateStateTimer.stop() as well to make sure the STM is not executed in this state.